### ISSUE TITLE
archive.extracted breakage due to 2016.3.0 upgrade from 2015.8.10

### ISSUE THREAD

**Issue by windoverwater** [saltstack/salt#34101]
### Description of Issue/Question

Hi

We upgraded from 2015.8 to 2016.3 and most of our file.archive states broke.  They apparently broke because we are using both user, group, and if_missing, and in most/all cases if_missing is pointing to a specific file.

Apparently, there is a non backwards compatible change in 2016.3 (as documented):  "Changed in version 2016.3.0: When used in combination with if_missing, ownership will only be enforced if if_missing is a directory"/

We had been using file.archive to place into existing directory structures additional sub footprints delivered to us as zip, tar, and war files.  And we were relying on the user and group switches so to be able to correctly set the file ownership.  Note that some of these zip files are being placed into pre existing (!) directories.

Is there any way to get back to the 2015.8 functionality of the file.archive function naturally setting the file ownership INDEPENDENT of whether if_missing is pointing to a directory or file?

In addition, even when we set if_missing to a directory, in our limited testing of this scenario the archive is still exploding with root ownership of the files (salt is running as root) - the user and group args are not being honored.  Is 2016.3 really supposed to behave as documented - which as we interpret, is if if_missing points to a directory, the user and group args are honored, while if if_missing points to a non-directory inode, user and group are not honored?  Is there a test that we can run to validate that behavior?
### Setup

Sorry - don't have one handy.  Yes this is lame, but a single file.archive state with a zip files of a small directory would be enough of a replication test.
### Steps to Reproduce Issue

For issue one, the question is about breaking backwards compatibility with 2015.8 documented behavior (as well as observed behavior).  

For issue two, we cannot seem to get the user and group arguments to be honored even as documented in 2016.3.
### Versions Report

$ sudo salt --versions-report
Salt Version:
           Salt: 2016.3.0

Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: 1.4.1
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.7.3
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: 0.20.2
           Mako: 0.3.4
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
         pygit2: Not Installed
         Python: 2.6.6 (r266:84292, May 22 2015, 08:34:51)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 14.5.0
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.0.5

System Versions:
           dist: redhat 6.8 Santiago
        machine: x86_64
        release: 2.6.32-504.23.4.el6.x86_64
         system: Linux
        version: Red Hat Enterprise Linux Server 6.8 Santiago

**Comment by Ch3LL** [comment]
@windoverwater I'm assuming you are refering to `archive.extracted` not `file.extracted`. looks like this behavior was introduced with this [PR](https://github.com/saltstack/salt/pull/32893)

ping @terminalmage looks like in your comments in the PR you do state that there might be difficulties to chmod an individual file hence the reason it needs to be a directory. do you think it could be possible at all to ensure that is_missing does not need to be a directory? Or maybe a workaround?

**Comment by windoverwater** [comment]
Apologies - yes I mean archive.extracted.  FWIIW 2016.3 would work for us if at the implementation level the implementation effectively did a chmod -R at the root directory of the archive.extract.  Yes we place extracted trees within other trees, but the order of extraction is _always_ from folder to subfolder (downwards) and never from subfolder to folder (upwards).  So I think a single effective 'chmod -R' would be ok for our specific use case.

In fact our quick fix has been to add an explicit cmd.run state that does that - it runs a 'chmod -R' after the extract.

**Comment by somenick** [comment]
We're also using a marker file pattern, although not in the same way as the reporter of the #32801 that resulted in the PR above.

**Comment by gladiatr72** [comment]
__bump__

`salt-minion 2016.3.3 (Boron)`

---

TL;DR

`archive.extracted` has run off the functionality rails.  If you are trying to get it to do something other than extract archives into a particular spot then you will find it probably not doing what you think it ought to be doing even though the docs leave one with the impression that it should be.

Use this state to extract archives.  It does that well.  The file/directory ownership and the `if_missing` arguments are not generally doing what they ought to in the contexts that most of us probably use it.  You can achieve your intended goal by making `archive.extracted` a target of a `file.exists` `onfail` requisite.  

Use a `file.directory` state to set the directory/file permissions as needed (this state function is also on the fancy side but everything is well integrated and functional).  If you don't want to run a permissions sweep on every state run, you can tack an `onchanges` requisite targeted at the `archive.extracted` state.

---
(less constructive rantiness starts here)

From the [docs](https://docs.saltstack.com/en/latest/ref/states/all/salt.states.archive.html#salt.states.archive.extracted): 

`name` = __Location where archive should be extracted__ 

`if_missing` = If specified, this path will be checked, and if it exists then the archive will not be extracted. This can be helpful if the archive extracts all files into a subfolder. This path can be either a directory or a file, so this option can also be used to check for a semaphore file and conditionally skip extraction.

__Changed in version 2016.3.0: When used in combination with either user or group, ownership will only be enforced when if_missing is a directory__

Ok.  Fine.  Whatever.  I'm sure there is a perfectly valid reason why this is.. erm.. the way it is.

But then there is [__this__](https://github.com/saltstack/salt/blame/v2016.3.3/salt/states/archive.py#L270) that assigns the value of `name` to the `if_missing` variable if `if_missing` is `None`.

`name` = __Location where archive should be extracted__ 

Ok, so, let's say `name` = '/tmp' because that is the location where the archive should be extracted.  It exists.  Or rather, if it doesn't exist, I want my state to blow up and let me know because `archive.extracted` knows nothing of `/tmp` or any other target archive extraction location _just like it should_.

Now let's say I've got an archive  that drops a directory named `/tmp/somethingreallylong-i_like_pie-and-chips_frogs-over-toads__babylon5-vs-startrek--omgwhatwastheguythatmadethisthingthinking` and I have a follow-up state that renames it to `/tmp/some_data`.  Now, `/tmp/some_data` is not (yet) a directory--it doesn't exist yet, right?  

So, no problem.  The directory gets renamed.  The next time this state runs, the archive isn't extracted again because it rightly identifies the `if_missing` value as a directory but..

Well, FFFFFFFIDDLESTICKS!   So, I had specified both `user` and `group`.  But the extracted archive's contents are owned by... uid `500`????  That sometimes happens when I've extracted tarballs with the `p` option but the user-of-origin doesn't exist on the local system.

```
vim salt/states/archive.py
```

:face_with_head_bandage: 

I don't understand this business of ignoring the user/group args if `if_missing` points at something that isn't a directory.

I don't understand the business of considering `tar` archives that don't include a directory at their root to be the `common` use.  I've been extracting tar archives for a Very Long Time.  The last time I ran into one that didn't have a subdirectory at its root was also the last time I extracted one from a (神聖的睾丸!) floppy disk!

I got caught *not* taking my oh-so-very-sage-advice earlier today which is how I ended up here.  In my defense, it was from recycling some states from a previous project that was probably running on early 2015 or 2014 salt masters.

`archive.extracted` is drek and is in desperate need of a re-write.

That's all.  I'm done now.

**Comment by terminalmage** [comment]
The ``archive.extracted`` state is not *just* for tar archives.** ZIP archives are also supported, and a single top-level directory is most definitely _not_ the "common" use for ZIP archives. ``if_missing`` was therefore likely initially introduced as an attempt to shoehorn idempotence onto a state that needs to support more that one kind of archive type.

Given the differences in use between tar and zip archives, it is probably a good idea to add separate states for extracting tar and zip archives, and deprecate ``archive.extracted``.

**Comment by terminalmage** [comment]
OK, we had some internal discussion and the decision has been made not to make separate states, but to change how the `archive.extracted` state works. Right now we decide what files and dirs have ownership managed based on `if_missing`, because we don't know the contents of the archive and if there are other files and dirs in the same parent directory, they would get ownership enforced as well if we just did it on the parent directory.

So, I wrote a function which lists the contents of a tar, zip, or rar archive, and we can use this information to determine what to manage ownership for. `if_missing` will become a simple check for the existence of a file/dir at that path.

For cases like the one in which the top level dir is renamed, an argument will be provided to specify a path or paths to recursively have their ownership managed.

**Comment by terminalmage** [comment]
Pull request #37368 has been merged, which significantly overhauls the ``archive.extracted`` state. Rather than re-state what changed, you can view [this diff](https://github.com/saltstack/salt/pull/37368/files#diff-8c5197ae1d9bebf29578f139e8cc8f25R139) from the pull request to see the updates to the release notes for the upcoming 2016.11.0 (Carbon) release.
